src/backend/access/transam/xact.c

```diff
@@ -6828,12 +6852,23 @@ XactLogCommitRecord(TimestampTz commit_time,
 		xl_origin.origin_timestamp = replorigin_session_origin_timestamp;
 	}
 
-	if (isDtxPrepared || isOnePhaseQE)
+	if (isDtxPrepared || isOnePhaseQE || (XLogLogicalInfoActive() && info == XLOG_XACT_COMMIT_PREPARED))
 	{
 		xl_xinfo.xinfo |= XACT_XINFO_HAS_DISTRIB;
 		xl_distrib.distrib_xid = getDistributedTransactionId();
 	}
 
+	/*
+	 * When QE commit prepared transaction, and the transaction is one-phase,
+	 * there will be no relative xlogs on master. 
+	 * We need to distinguish this special situation when we do logical decoding.
+	 */
+	if (XLogLogicalInfoActive() && info == XLOG_XACT_COMMIT && isOnePhaseQE)
+	{
+		xl_xinfo.xinfo |= XACT_XINFO_HAS_ISONEPHASE;
+		xl_isonephase.isonephase = 1;
```

Comment:
I think you should keep one naming style. for example "xl_is_one_phase" or "isOnePhase"